x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/classification/Precision.java

```diff
@@ -0,0 +1,321 @@
+import org.elasticsearch.xpack.core.ml.dataframe.evaluation.EvaluationMetric;
+import org.elasticsearch.xpack.core.ml.dataframe.evaluation.EvaluationMetricResult;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.xpack.core.ml.dataframe.evaluation.MlEvaluationNamedXContentProvider.registeredMetricName;
+
+/**
+ * {@link Precision} is a metric that answers the question:
+ *   "What fraction of documents classified as X actually belongs to X?"
+ * for any given class X
+ *
+ * equation: precision(X) = TP(X) / (TP(X) + FP(X))
+ * where: TP(X) - number of true positives wrt X
+ *        FP(X) - number of false positives wrt X
+ */
+public class Precision implements EvaluationMetric {
+
+    public static final ParseField NAME = new ParseField("precision");
+
+    private static final String PAINLESS_TEMPLATE = "doc[''{0}''].value == doc[''{1}''].value";
+    private static final String AGG_NAME_PREFIX = "classification_precision_";
+    static final String ACTUAL_CLASSES_NAMES_AGG_NAME = AGG_NAME_PREFIX + "by_actual_class";
+    static final String BY_PREDICTED_CLASS_AGG_NAME = AGG_NAME_PREFIX + "by_predicted_class";
+    static final String PER_PREDICTED_CLASS_PRECISION_AGG_NAME = AGG_NAME_PREFIX + "per_predicted_class_precision";
+    static final String AVG_PRECISION_AGG_NAME = AGG_NAME_PREFIX + "avg_precision";
+
+    private static String buildScript(Object...args) {
+        return new MessageFormat(PAINLESS_TEMPLATE, Locale.ROOT).format(args);
+    }
+
+    private static final ObjectParser<Precision, Void> PARSER = new ObjectParser<>(NAME.getPreferredName(), true, Precision::new);
+
+    public static Precision fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    private List<String> topActualClassNames;
+    private EvaluationMetricResult result;
+
+    public Precision() {}
+
+    public Precision(StreamInput in) throws IOException {}
+
+    @Override
+    public String getWriteableName() {
+        return registeredMetricName(Classification.NAME, NAME);
+    }
+
+    @Override
+    public String getName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    public final Tuple<List<AggregationBuilder>, List<PipelineAggregationBuilder>> aggs(String actualField, String predictedField) {
+        if (topActualClassNames == null) {  // This is step 1
+            return Tuple.tuple(
+                List.of(
+                    AggregationBuilders.terms(ACTUAL_CLASSES_NAMES_AGG_NAME)
+                        .field(actualField)
```

Comment:
I've just added `size` parameter to `Precision` and `Recall` metrics.
Also, I've added `other_class_count` parameter to `Precision.Result` and `Recall.Result` so that the user can tell if the result is complete.

Reply:
> I've just added size parameter to Precision and Recall metrics.
Also, I've added other_class_count parameter to Precision.Result and Recall.Result so that the user can tell if the result is complete.

I reverted this change and implemented max cardinality enforcement as discussed. PTAL